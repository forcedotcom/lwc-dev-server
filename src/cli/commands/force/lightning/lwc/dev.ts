import { flags, SfdxCommand } from '@salesforce/command';
import { Messages, SfdxError } from '@salesforce/core';
import { AnyJson, JsonArray, JsonMap } from '@salesforce/ts-types';
import * as http from 'http';
import Project from '../../../../../common/Project';
import SfdxConfiguration from '../../../../../user/SfdxConfiguration';
import LocalDevServer from '../../../../../server/LocalDevServer';
import debugLogger from 'debug';

const debug = debugLogger('localdevserver');

// Initialize Messages with the current plugin directory
Messages.importMessagesDirectory(__dirname);

// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
// or any library that is using the messages framework can also be loaded this way.
const messages = Messages.loadMessages('lwc-dev-server', 'dev');

export default class Dev extends SfdxCommand {
    public static description = messages.getMessage('commandDescription');

    public static examples = [
        `$ sfdx force:lighting:lwc:dev --open myComponent`
    ];

    public static args = [{ open: 'file', name: 'lwc-dev' }];

    protected static flagsConfig = {
        open: flags.string({
            char: 'o',
            description: messages.getMessage('openFlagDescription')
        }),
        port: flags.integer({
            char: 'p',
            description: messages.getMessage('portFlagDescription')
        })
    };

    // Comment this out if your command does not require an org username
    protected static requiresUsername = true;

    // Comment this out if your command does not support a hub org username
    protected static supportsDevhubUsername = true;

    // Set this to true if your command requires a project workspace; 'requiresProject' is false by default
    protected static requiresProject = true;

    public async run(): Promise<AnyJson> {
        const componentName = this.flags.open;
        const port = this.flags.port || 3333;

        if (!this.org) {
            this.ux.log('org was undefined, an org is required.');
            return { org: typeof this.org };
        }

        if (!this.project) {
            this.ux.log(
                'project is undefined. this must be run from a sfdx worksapce'
            );
            return { project: typeof this.project };
        }

        // this.org is guaranteed because requiresUsername=true, as opposed to supportsUsername
        const conn = this.org.getConnection();

        // Highest level API is always last
        const api_version = await conn.retrieveMaxApiVersion();

        debug(
            `You appear to be running on a Salesforce instance that can support up to API level ${api_version}`
        );

        if (componentName) {
            debug(`You wanted to open this component: ${componentName}`);
        }

        // TODO resolve location of `componentName`, ensure directory structure is imported / compiled

        // TODO check if it's already running on the port first

        // custom onProxyReq function to inject into Talon's proxy
        // this will insert the Authorization header to have the requests be authenticated
        const onProxyReq = function(
            proxyReq: http.ClientRequest,
            req: http.IncomingMessage,
            res: http.ServerResponse
        ) {
            proxyReq.setHeader('Authorization', `Bearer ${conn.accessToken}`);
            // req.headers.Cookie = `sid=${sid_cookie}`;
        };

        const project = new Project(this.project.getPath());

        const sfdxConfiguration = new SfdxConfiguration(project);
        sfdxConfiguration.api_version = api_version;
        sfdxConfiguration.endpoint = conn.instanceUrl;
        sfdxConfiguration.onProxyReq = onProxyReq;
        sfdxConfiguration.port = port;
        sfdxConfiguration.namespace = <string>(
            (await this.project.resolveProjectConfig()).namespace
        );

        project.sfdxConfiguration = sfdxConfiguration;

        const retValue = {
            orgId: this.org.getOrgId(),
            api_version: sfdxConfiguration.api_version,
            endpoint: sfdxConfiguration.endpoint,
            onProxyReq: JSON.stringify(sfdxConfiguration.onProxyReq),
            componentName,
            port,
            token: conn.accessToken
        };
        debug(JSON.stringify(retValue));

        // Start local dev server
        // TODO pass in component to open & open browser
        new LocalDevServer().start(project);

        return retValue;
    }
}
