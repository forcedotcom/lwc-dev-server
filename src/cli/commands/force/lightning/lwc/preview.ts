#!/usr/bin/env ts-node
import { flags, SfdxCommand } from '@salesforce/command';
import { Messages, SfdxError } from '@salesforce/core';
import { CommandLineUtils } from '../../../../../common/Common';
import { Launcher } from '../../../../../common/Requirements';
import { LaunchIOS } from '../../../../../common/LaunchIOS';
import { LaunchAndroid } from '../../../../../common/LaunchAndroid';
import { DefaultLWCMobileConfig } from '../../../../../common/Config';

// Initialize Messages with the current plugin directory
Messages.importMessagesDirectory(__dirname);

// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
// or any library that is using the messages framework can also be loaded this way.
const messages = Messages.loadMessages('@salesforce/lwc-dev-server', 'preview');

export default class Preview extends SfdxCommand {
    public static description = messages.getMessage('commandDescription');

    public static examples = [
        `$ sfdx force:lightning:lwc:preview -p iOS -t LWCSim2 -f http://localhost:3333`,
        `$ sfdx force:lightning:lwc:preview -p Android -t LWCEmu2 -f http://localhost:3333`
    ];

    public static args = [{ name: 'file' }];

    protected static flagsConfig = {
        // flag with a value (-n, --name=VALUE)
        platform: flags.string({
            char: 'p',
            description: messages.getMessage('platformFlagDescription')
        }),
        path: flags.string({
            char: 'f',
            description: messages.getMessage('pathFlagDescription')
        }),
        target: flags.string({
            char: 't',
            description: messages.getMessage('targetFlagDescription')
        })
    };

    // Comment this out if your command does not require an org username
    protected static requiresUsername = false;

    // Comment this out if your command does not support a hub org username
    protected static supportsDevhubUsername = false;

    // Set this to true if your command requires a project workspace; 'requiresProject' is false by default
    protected static requiresProject = false;

    public async run(): Promise<any> {
        let launcher: Launcher;
        let url: string = '';

        if (this.flags.platform === undefined) {
            return new Promise<any>((resolve, reject) => reject(false));
        }
        //FIXME: Invoke setup and use detected settings!
        if (CommandLineUtils.platformFlagIsIOS(this.flags.platform)) {
            launcher = new LaunchIOS(this.flags.target, 'iPhone-X', 'iOS-13-3');
            url = this.flags.path;
        } else if (
            CommandLineUtils.platformFlagIsAndroid(this.flags.platform)
        ) {
            launcher = new LaunchAndroid(
                this.flags.target,
                'pixel',
                'android-28'
            );
            url = this.flags.path;
        }

        // if (this.flags.platform.toLowerCase() == "ios") {
        //   launcher = new LaunchIOS(this.flags.target, "iPhone-X","iOS-13-3");
        //   url = this.flags.path;
        // } else if (this.flags.platform.toLowerCase() == "android") {
        //   launcher = new LaunchAndroid(this.flags.target,"pixel","android-28");
        //   url = this.flags.path;
        // }
        // return launcher.launchNativeBrowser(url);
        return new Promise<any>((resolve, reject) => {
            if (launcher == null) {
                reject('Could not launch target');
                return;
            }
            resolve(launcher.launchNativeBrowser(url));
        });
    }
}
