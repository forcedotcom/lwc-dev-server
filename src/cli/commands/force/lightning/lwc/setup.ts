#!/usr/bin/env ts-node

import { flags, SfdxCommand } from '@salesforce/command';
import { Messages, SfdxError } from '@salesforce/core';
import { CommandLineUtils } from '../../../../../common/Common';
import {
    RequirementList,
    SetupTestResult
} from '../../../../../common/Requirements';
import { AndroidEnvironmentSetup } from '../../../../../common/AndroidEnvironmentSetup';
import { IOSEnvironmentSetup } from '../../../../../common/IOSEnvironmentSetup';

// Initialize Messages with the current plugin directory
Messages.importMessagesDirectory(__dirname);

// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
// or any library that is using the messages framework can also be loaded this way.
const messages = Messages.loadMessages('@salesforce/lwc-dev-server', 'preview');

export default class Setup extends SfdxCommand {
    public static description = messages.getMessage('commandDescription');

    public static examples = [
        `$ sfdx force:lightning:lwc:setup -p iOS`,
        `$ sfdx force:lightning:lwc:setup -p Android`
    ];

    protected static flagsConfig = {
        // flag with a value (-n, --name=VALUE)
        platform: flags.string({
            char: 'p',
            description: messages.getMessage('platformFlagDescription')
        })
    };

    public async run(): Promise<SetupTestResult> {
        let setup: RequirementList;
        //let setupResult: SetupTestResult = new  SetupTestResult();

        if (this.flags.platform === undefined) {
            return new Promise((resolve, reject) => {
                reject({ hasMetAllRequirements: false, tests: [] });
            });
        }

        if (CommandLineUtils.platformFlagIsIOS(this.flags.platform)) {
            setup = new IOSEnvironmentSetup();
        } else {
            setup = new AndroidEnvironmentSetup();
        }
        return setup.executeSetup();
    }
}
