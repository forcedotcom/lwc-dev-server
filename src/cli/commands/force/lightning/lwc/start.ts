import { flags, SfdxCommand } from '@salesforce/command';
import { Messages } from '@salesforce/core';
import { AnyJson } from '@salesforce/ts-types';
import * as http from 'http';
import Project from '../../../../../common/Project';
import LocalDevServer from '../../../../../server/LocalDevServer';
import debugLogger from 'debug';
import colors from 'colors';

const debug = debugLogger('localdevserver');

// Initialize Messages with the current plugin directory
Messages.importMessagesDirectory(__dirname);

// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
// or any library that is using the messages framework can also be loaded this way.
const messages = Messages.loadMessages('@salesforce/lwc-dev-server', 'start');

export default class Start extends SfdxCommand {
    public static description = messages.getMessage('commandDescription');

    public static examples = [
        `$ sfdx force:lighting:lwc:start`,
        `$ sfdx force:lighting:lwc:start --port 3000`
    ];

    public static args = [];

    protected static flagsConfig = {
        port: flags.integer({
            char: 'p',
            description: messages.getMessage('portFlagDescription')
        })
    };

    // Comment this out if your command does not support specifying an org username
    protected static supportsUsername = true;

    // Comment this out if your command does not support a dev hub org username
    protected static supportsDevhubUsername = true;

    // Set this to true if your command requires a project workspace; 'requiresProject' is false by default
    protected static requiresProject = true;

    public async run(): Promise<AnyJson> {
        const defaultdevhubusername = this.configAggregator.getPropertyValue(
            'defaultdevhubusername'
        ) as string;

        let port: number;
        if (this.flags.port !== undefined && this.flags.port !== null) {
            port = this.flags.port;
        } else {
            port = 3333;
        }

        if (!this.org) {
            // This you DO need.
            // We require this right now for proxying and api version.
            // If we do not have an org, we can still function.
            //  - We should disable the proxying and then allow specification of the version
            //    if you don't authenticate.
            //

            const targetusername = this.flags.targetusername;
            if (targetusername) {
                this.reportStatus(
                    colors.green(defaultdevhubusername),
                    colors.red(
                        `${targetusername} - ${messages.getMessage(
                            'error:invalidscratchorgusername'
                        )}`
                    )
                );
            } else {
                const configuredusername = this.configAggregator.getPropertyValue(
                    'defaultusername'
                );
                this.reportStatus(
                    colors.green(defaultdevhubusername),
                    colors.red(
                        `${configuredusername} - ${messages.getMessage(
                            'error:noscratchorg'
                        )}`
                    )
                );
            }

            return { org: typeof this.org };
        }

        // Sfdx validates this before we have a chance to, this appears to be
        // a "just in case" condition so reporting the same error they do.
        if (!this.project) {
            this.ux.error(messages.getMessage('error:noproject'));
            return { project: typeof this.project };
        }

        this.ux.log(colors.gray(messages.getMessage('legal:cliusage')));

        // this.org is guaranteed because requiresUsername=true, as opposed to supportsUsername
        const conn = this.org.getConnection();

        // Highest level API is always last
        const api_version = await conn.retrieveMaxApiVersion();

        const orgusername = this.org.getUsername() || '';
        try {
            // currently something in sfdx is resulting in an unhandled
            // promise rejection, instead of surfacing that rejection
            // through the returned promise. @W-6723813
            process.on('unhandledRejection', reason => {
                if (
                    reason &&
                    // @ts-ignore
                    reason.name &&
                    // @ts-ignore
                    reason.name === 'StatusCodeError'
                ) {
                    // ignore unhandled rejects during below refresh call
                    debug(`unhandledPromiseRejection: ${reason}`);
                } else {
                    this.ux.error(`unhandledPromiseRejection: ${reason}`);
                }
            });
            await this.org.refreshAuth();
        } catch (err) {
            this.reportStatus(
                colors.green(defaultdevhubusername),
                colors.red(
                    `${orgusername} - ${messages.getMessage(
                        'error:inactivescratchorg'
                    )}`
                ),
                colors.green(api_version)
            );
            return { error: err };
        }

        this.reportStatus(
            colors.green(defaultdevhubusername),
            colors.green(orgusername),
            colors.green(api_version)
        );

        const accessToken = conn.accessToken;

        // custom onProxyReq function to inject into Talon's proxy
        // this will insert the Authorization header to have the requests be authenticated
        const onProxyReq = function(
            proxyReq: http.ClientRequest,
            req: http.IncomingMessage,
            res: http.ServerResponse
        ) {
            proxyReq.setHeader('Authorization', `Bearer ${accessToken}`);
        };

        const project = new Project(this.project.getPath());

        project.configuration.api_version = api_version;
        project.configuration.endpoint = conn.instanceUrl;
        project.configuration.onProxyReq = onProxyReq;
        project.configuration.port = port;
        project.configuration.namespace = <string>(
            (await this.project.resolveProjectConfig()).namespace
        );

        const retValue = {
            orgId: this.org.getOrgId(),
            api_version: project.configuration.api_version,
            endpoint: project.configuration.endpoint,
            onProxyReq: JSON.stringify(project.configuration.onProxyReq),
            port,
            token: accessToken
        };
        debug(JSON.stringify({ ...retValue, token: undefined }));

        // Start local dev server
        new LocalDevServer().start(project, conn);

        return retValue;
    }

    private reportStatus(
        devHubOrg: string,
        scratchOrg: string,
        apiVersion?: string
    ) {
        if (apiVersion) {
            this.ux.log(`\
Starting LWC Local Development.
    Dev Hub Org: ${devHubOrg}
    Scratch Org: ${scratchOrg}
    Api Version: ${apiVersion}\
`);
        } else {
            this.ux.log(`\
Starting LWC Local Development.
    Dev Hub Org: ${devHubOrg}
    Scratch Org: ${scratchOrg}\
`);
        }
    }
}
