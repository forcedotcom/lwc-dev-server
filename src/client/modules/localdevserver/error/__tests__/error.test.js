import { createElement } from 'lwc';
import ErrorStacks from 'localdevserver/error';

function createComponentUnderTest(props) {
    const el = createElement('localdevserver-error', {
        is: ErrorStacks
    });
    Object.assign(el, props);
    document.body.appendChild(el);
    return el;
}

describe('localdevserver-error-stacks', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        jest.resetModules();
    });

    afterEach(() => {
        // The jsdom instance is shared across test cases in a single file so reset the DOM
        while (document.body.firstChild) {
            document.body.removeChild(document.body.firstChild);
        }
    });

    it('should render no error', () => {
        const componentElement = createComponentUnderTest();
        expect(componentElement).toMatchSnapshot();
    });

    it('should render simple errors', () => {
        const componentElement = createComponentUnderTest({
            error: {
                message: 'simple error message'
            }
        });
        const errorMsg = componentElement.shadowRoot.querySelector(
            '.error-message'
        ).textContent;
        expect(errorMsg).toMatchSnapshot();
    });

    it('should trigger fetch call', async () => {
        const mockComponentDidMount = {
            ok: true,
            json: () =>
                Promise.resolve({
                    errors: [
                        {
                            location: { line: 1, column: 2 },
                            message:
                                'SyntaxError: /Users/sample/ebikes-lwc/force-app/main/default/lwc/productTile/productTile.js: LWC1100: Invalid decorator usage. Supported decorators (api, wire, track) should be imported from "lwc"',
                            code: `
                            7 | export default class ProductTile extends LightningElement {
                            8 |     /** Whether the tile is draggable. */
                            >  9 |     @api draggable;
                                |     ^
                            10 |
                            11 |     _product;`,
                            filename: 'productTile.js'
                        }
                    ]
                })
        };

        global.fetch = jest
            .fn()
            .mockReturnValue(Promise.resolve(mockComponentDidMount));

        const componentElement = createComponentUnderTest({
            error: {
                specifier: 'c/productTile'
            }
        });
        expect(global.fetch).toHaveBeenCalledTimes(1);
        /* 
        NOTE: wasn't able to run the test and wait for the fetch mock to finish running before
        asserting, will keep looking for how to accomplish that
        const href = componentElement.shadowRoot
            .querySelector('a')
            .getAttribute('href');

        expect(href).toBe(
            `vscode://file${error.filename}:${error.location.line}:${error.location.column}`
        ); */
    });
});
