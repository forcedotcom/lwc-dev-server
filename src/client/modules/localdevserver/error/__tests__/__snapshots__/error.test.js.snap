// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`localdevserver-error-stacks should render no error 1`] = `
<localdevserver-error>
  #shadow-root(open)
    <section
      aria-modal="true"
      class="slds-modal slds-fade-in-open slds-modal_large"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="slds-modal__container"
      >
        <header
          class="slds-modal__header"
        >
          <h2
            id="[shadow:guid]"
          >
            <img
              class="error-image"
              src="/assets/localdev/images/buggy.svg"
            />
          </h2>
        </header>
        <div
          class="slds-modal__content slds-p-around_medium"
        >
          <h1
            class="error-message slds-text-heading_large"
          />
          <h2
            class="error-location slds-text-heading_small"
          />
          <localdevserver-code-highlighter
            class="error-code"
          >
            #shadow-root(open)
              <div />
          </localdevserver-code-highlighter>
          <localdevserver-error-stacks
            class="error-stacks"
          >
            #shadow-root(open)
              <div
                class="stack-message slds-text-heading_small"
              >
                <div
                  class="button-container"
                >
                  <div
                    class="line"
                  />
                  <div
                    class="collapse-container"
                  >
                    <div
                      class="collapse"
                    >
                      0 stack frames collapsed
                    </div>
                  </div>
                </div>
              </div>
          </localdevserver-error-stacks>
          <div
            class="info-note slds-text-heading_small"
          >
            This screen is visible only in development. It will not appear if the app crashes in production. Open your browser's developer console to further inspect this error.
          </div>
        </div>
      </div>
    </section>
    <div
      class="slds-backdrop slds-backdrop_open"
    />
</localdevserver-error>
`;

exports[`localdevserver-error-stacks should render simple errors 1`] = `"simple error message"`;
